File: N_Queens.java
public class NQueens {
    private static final int N = 8;

    public static void main(String[] args) {
        solveNQueens(N);
    }

    private static void solveNQueens(int n) {
        char[][] board = new char[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i][j] = '.';
            }
        }
        solveNQueensUtil(board, 0);
    }

    private static void solveNQueensUtil(char[][] board, int row) {
        if (row == board.length) {
            printBoard(board);
            return;
        }

        for (int col = 0; col < board.length; col++) {
            if (isSafe(board, row, col)) {
                board[row][col] = 'Q'; // Place queen
                solveNQueensUtil(board, row + 1); // Recur to place the rest
                board[row][col] = '.'; // Backtrack
            }
        }
    }

    private static boolean isSafe(char[][] board, int row, int col) {
        // Check this column on upper side
        for (int i = 0; i < row; i++) {
            if (board[i][col] == 'Q') {
                return false;
            }
        }

        // Check upper diagonal on left side
        for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        // Check upper diagonal on right side
        for (int i = row, j = col; i >= 0 && j < board.length; i--, j++) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        return true;
    }

    private static void printBoard(char[][] board) {
        for (char[] row : board) {
            for (char c : row) {
                System.out.print(c + " ");
            }
            System.out.println();
        }
        System.out.println();
    }
}

File: N_Queens.rb
class NQueens
  N = 8

  def self.solve_n_queens
    board = Array.new(N) { Array.new(N, '.') }
    solve_n_queens_util(board, 0)
  end

  def self.solve_n_queens_util(board, row)
    if row == board.length
      print_board(board)
      return
    end

    (0...board.length).each do |col|
      if is_safe(board, row, col)
        board[row][col] = 'Q' # Place queen
        solve_n_queens_util(board, row + 1) # Recur to place the rest
        board[row][col] = '.' # Backtrack
      end
    end
  end

  def self.is_safe(board, row, col)
    # Check this column on upper side
    (0...row).each do |i|
      return false if board[i][col] == 'Q'
    end

    # Check upper diagonal on left side
    (row).downto(0) do |i|
      j = col - (row - i)
      return false if j >= 0 && board[i][j] == 'Q'
    end

    # Check upper diagonal on right side
    (row).downto(0) do |i|
      j = col + (row - i)
      return false if j < board.length && board[i][j] == 'Q'
    end

    true
  end

  def self.print_board(board)
    board.each do |row|
      puts row.join(' ')
    end
    puts
  end
end

NQueens.solve_n_queens
